model/utils/price_impact_valuator.py

```diff
@@ -0,0 +1,92 @@
+"""
+import numpy as np
+
+from experiments import simulation_configuration
+from model.types import ImpactDelay, PriceImpact
+
+
+class PriceImpactValuator():
+    """
+    This class evaluates the price impact of trades with CEX / general off-chain market
+    """
+
+    def __init__(self, impacted_assets, sample_size):
+        self.supply_changes = {asset.split('_')[0]: np.zeros(
+            simulation_configuration.BLOCKS_PER_TIMESTEP
+            * simulation_configuration.TIMESTEPS
+            + 1) for asset in impacted_assets}
+        self.sample_size = sample_size
+        self.price_impact_model = PriceImpact.ROOT_QUANTITY
+
+    def price_impact(
+        self,
+        floating_supply,
+        pre_floating_supply,
+        current_step,
+        market_prices,
+        params
+    ):
+        """
+        This functions evaluates price impact of supply changes
+        """
+        block_supply_change = {
+            ccy: supply - pre_floating_supply[ccy]
+            for ccy, supply in floating_supply.items()
+        }
+        self.impact_delay(block_supply_change, current_step, params)
+
+        price_impact = {}
+        prices = {}
+
+        for asset in params['impacted_assets']:
+            asset_1, _ = asset.split('_')
+            if asset_1 == 'usd':
+                raise Exception(f'Incorrect quoting convention for {asset}')
+            variance_daily = params["variance_market_price"][asset] / 365
+            average_daily_volume = params["average_daily_volume"][asset]
+            price_impact[asset] = self.price_impact_function(self.price_impact_model)(
+                self.supply_changes[asset_1][current_step],
+                variance_daily,
+                average_daily_volume,
+            )
+            prices[asset] = market_prices[asset] + price_impact[asset]
+
+        return prices
+
+    def impact_delay(
+        self, block_supply_change, current_step, params
+    ):
+        """
+        This function distributes / delays supply changes across future timesteps
+        """
+        impact_delay = params['impact_delay']
+        for ccy in block_supply_change:
+            if impact_delay['model'] == ImpactDelay.INSTANT:
+                self.supply_changes[ccy][current_step] += block_supply_change[ccy]
+
+            elif impact_delay['model'] == ImpactDelay.NBLOCKS:
+                delay = impact_delay['param_1']
+                for block in range(current_step, current_step +
+                                   min(delay, self.sample_size % current_step)):
+                    self.supply_changes[ccy][block] += (1 /
+                                                        min(delay,
+                                                            self.sample_size %
+                                                            current_step) *
+                                                        block_supply_change[ccy])
+
+    # pylint: disable=no-self-use
```

Comment:
had that a couple of times too, what do you think is better, making such functions without self use staticmethods or keep them as 'object' functions? Would be curious about @bowd's view to

Reply:
I guess, it could be stored in a utils file. As it is not too large I kept it in the class. staticmethod would def be an option

Reply:
If the function is useful outside of this class and want to be kept tied to it, for example by needing to call ...
```python
PriceImpactValuator.price_impact_function(...)
```
... somewhere else in the codebase, then I would make it a `@staticmethod`.

Otherwise if this is just an utility function I would move it outside of the class, but in the same file and make it a "private" function:
```python
def __price_impact_function__(...):
   # ... 
```

And call it from the class.